File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void helper(int i, int n, vector<int>&nums, vector<int>&ds, vector<vector<int>>&ans){
        if(i == n){
            ans.push_back(ds);
            return;
        }

        ds.push_back(nums[i]);
        helper(i+1, n, nums, ds, ans);
        ds.pop_back();
        helper(i+1, n, nums, ds, ans);
    }   

    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        vector<int> ds;

        helper(0, n, nums, ds, ans);
        return ans;
        
    }
};

File: 0078-subsets/0078-subsets.py
class Solution:
    def helper(self, ind, nums, ds, ans):
        if ind == len(nums):
            ans.append(ds[:])
            return

        ds.append(nums[ind])
        self.helper(ind + 1, nums, ds, ans)
        ds.pop()
        self.helper(ind + 1, nums, ds, ans)
        
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        ds = []
        self.helper(0, nums, ds, ans)
        return ans